parser.py

```diff
@@ -121,35 +158,87 @@ def __init__(self, coeff, op):
         self.coeff = coeff
         self.op = op
 
-    def __str__(self):
-        return "{} {}".format(self.coeff, self.op)
-
     def __repr__(self):
-        return self.__str__()
+        return "{} {}".format(self.coeff, self.op)
 
 
 class Declaration:
     def __init__(self, decl_type, **kwargs):
         self.name = decl_type
         self.kwargs = kwargs
+        self.opaque = False
```

Comment:
I wasn't _really_ sure about this, whether opaque declarations should be separate objects, or simply a flag.

Something else I couldn't work out from the spec: whether a non-opaque declaration with a missing `goplist` should raise an error (I think it should, you then have to be explicit with the `opaque` keyword)

Reply:
the opaque gates do not have to define their constituent operations. It's a keyword in the spec that allows to bypass a specific definition (assuming the final consumer still knows what that gate is)

Reply:
Yep. So we should raise an exception if a gate declaration is given with no constituent ops, and no `opaque` keyword